### ISSUE TITLE
After window resized to different aspect ratio content displayed incorrectly

### ISSUE THREAD

**Issue by MarkCallow** [xbmc/xbmc#25875]
### Describe the bug
Here is a clear and concise description of what the problem is:

When resizing a window on macOS in a way which changes its aspect ratio, the GUI and any played video is displayed with an incorrect aspect ratio.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

I expect circles to remain circles and images and text to be displayed with the same aspect ratio after the size change.

## Actual Behavior

The aspect ratio (i.e. the projection matrix) used to display the graphical content of the window is wrong. This is most easily seen by moving to the 3 controls at the top of the left pane of the Kodi's main screen. The locate highlight should be a circle. Instead it is clearly oval.

You can also see that the displayed thumbnails are squashed, stretched or in some cases cropped and text is also squashed or stretched.

## Possible Fix

The projection matrix is pretty obviously being set incorrectly after a window resize.

### To Reproduce

1. Start Kodi
2. Resize the window by stretching a side, top or bottom of the window, or by entering full screen if the window has a different aspect ratio than the display. Do not stretch diagonally as this will not change the aspect ratio thus not reveal the problem.
3.

### Debuglog

The debuglog can be found here:

https://paste.kodi.tv/ciqoheqodi.kodi

### Screenshots 
Here are some links or screenshots to help explain the problem:

Screenshot at startup with power button highlighted.
<img width="1757" alt="Screenshot 2024-10-25 at 17 07 35" src="https://github.com/user-attachments/assets/8bf849ca-0a20-4082-9673-5c88e7e7943f">

Screenshot after entering full screen. This reveals the problem because the startup window size had a different aspect ratio than the screen. This shot shows a clearly oval highlight on the power button and distorted text and shrunk CD covers.
<img width="1280" alt="Screenshot 2024-10-25 at 17 07 54" src="https://github.com/user-attachments/assets/592bef97-0915-4fb2-b5ea-47c3ce771cc6">

If you now exit Kodi and restart, it will start up full screen and the content is now correctly displayed.
<img width="1280" alt="Screenshot 2024-10-25 at 17 08 56" src="https://github.com/user-attachments/assets/80a17033-067f-4629-beac-bff825d209a8">

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [x] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: macOS Sonoma 14.7
 - Kodi version: 21.0 and 21.1

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by enen92** [comment]
This is something I've been trying to solve for quite a while but with little success. @sarbes do you have any idea what might need to be reset to adjust to the new aspect ratio?
This affects all GL platforms not only macOS

**Comment by MarkCallow** [comment]
When a window is resized you have to update the projection matrix **and** the GL viewport to match the new size. You must make sure you have responded to the last resize event when the window size is being changed interactively so you are using the final size. If you are using scissor you will need to update that too.

**Comment by enen92** [comment]
@MarkCallow if it helps, whenever you resize the window the rendersystem reset method is called: https://github.com/xbmc/xbmc/blob/68effd58e050d8483c2de7d4c006e855c8011cb7/xbmc/rendering/gl/RenderSystemGL.cpp#L157
It seems to me that it already does what you describe but this is not really my field. Maybe you can spot what may be missing.
cheers

**Comment by Hitcher** [comment]
On Windows `ReloadSkin()` corrects it.

**Comment by sarbes** [comment]
I don't think that the matrix/viewport is wrong, we would see more major issues otherwise.

Given that Kodi is a fullscreen application and I don't see the need to change the window size, I have zero interests working on this.

**Comment by MarkCallow** [comment]
> I don't think that the matrix/viewport is wrong, we would see more major issues otherwise.

If the viewport is wrong, perhaps. If the projection matrix is wrong you will see exactly the issues that are being seen.

> Given that Kodi is a fullscreen application and I don't see the need to change the window size, I have zero interests working on this.

I have 2 displays on my computer of different aspect ratios and resolutions. This problem shows up when I move Kodi's full screen window from one display to another as was mentioned in #25697. Also I often find myself running Kodi in less than full screen on both my computer and my home theatre PC. Please do not ignore this bug.

**Comment by MarkCallow** [comment]
> It seems to me that it already does what you describe but this is not really my field. Maybe you can spot what may be missing.

At first glance nothing stands out. However I have been unable to find where the resize event is handled. The closest I have come is a ResizeWindow function but I failed to find the caller. I need to the whole path from event receipt.

**Comment by Hitcher** [comment]
In the meantime I suggest you try mapping `ReloadSkin()` to a key as that should fix it.

**Comment by DaVukovic** [comment]
@enen92 couldn't we reload the skin whenever the window size has changed?

**Comment by MarkCallow** [comment]
> @enen92 couldn't we reload the skin whenever the window size has changed?

I realize now that Kodi is using an orthographic projection not perspective. Thus it _is_ necessary to relayout the UI to the new aspect ratio when the window size changes. Calling `ReloadSkin()` when handling a resize event should fix it. @enen92 please do this.

@Hitcher how to you map ReloadSkin() to a key?

**Comment by Hitcher** [comment]
https://kodi.wiki/view/Keymap#Customizing_keymaps_through_the_GUI

**Comment by MarkCallow** [comment]
Given that I cannot use Kodi full-screen, because then moving the selection (using the arrow keys) turns off the HDMI display and sometimes messes up the color on the TV screen (not just the Kodi output)*, please, please, please fix this. @Hitcher I was never able to figure out how to map a key to `ReloadSkin()`. The ActionIDs linked from https://kodi.wiki/view/Keymap#Customizing_keymaps_through_the_GUI do not include `ReloadSkin`.

\* This has been the case since I switched to an m4 Mac Mini from a 2014 x64 Mac Mini. That switch was due to Nexus (see #25067) rendering my old device unusable. The disabling of the display seems related to my TV as I tried the new Mac Mini on a monitor with an HDMI input and the problem did not occur. Given that the TV is the only display connected to the Mac Mini, which is in my lounge, setting up an environment for debugging is difficult and I have not been able to gather enough information to make a useful bug report.

**Comment by MarkCallow** [comment]
I have created #28229 about the screen off issue.
